gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,307 @@
+            children.add(new OctreeNode(midx, midy, bounds.min.z, bounds.max.x, bounds.max.y, midz, deeperLevel));
+            children.add(new OctreeNode(bounds.min.x, midy, bounds.min.z, midx, bounds.max.y, midz, deeperLevel));
+            children.add(new OctreeNode(bounds.min.x, bounds.min.y, midz, midx, midy, bounds.max.z, deeperLevel));
+            children.add(new OctreeNode(midx, bounds.min.y, midz, bounds.max.x, midy, bounds.max.z, deeperLevel));
+            children.add(new OctreeNode(midx, bounds.min.y, bounds.min.z, bounds.max.x, midy, midz, deeperLevel));
+            children.add(new OctreeNode(bounds.min.x, bounds.min.y, bounds.min.z, midx, midy, midz, deeperLevel));
+
+            ObjectSet<T> geometries = this.geometries;
+            this.geometries = null;
+            // Move geometries from parent to children
+            for (OctreeNode child : children) {
+                for (T geometry : geometries) {
+                    child.add(geometry, maxItemsPerNode, collider);
+                }
+            }
+        }
+
+        public boolean contains (Collider<T> boundingBoxHandler, T geometry) {
+            return boundingBoxHandler.intersects(bounds, geometry);
+        }
+
+        public void add (T geometry, int maxItemsPerNode, Collider<T> collider) {
+            // If is not leaf, check children
+            if (!isLeaf()) {
+                for (OctreeNode child : children) {
+                    child.add(geometry, maxItemsPerNode, collider);
+                }
+                return;
+            }
+
+            if (!collider.intersects(bounds, geometry)) {
+                return;
+            }
+
+            addGeometry(geometry);
+            if (geometries.size > maxItemsPerNode && level > 0) {
+                split(maxItemsPerNode, collider);
+            }
+        }
+
+        public void remove (T aabb) {
+            if (isLeaf()) {
+                if (geometries != null) {
+                    geometries.remove(aabb);
+                }
+            } else {
+                for (OctreeNode node : children) {
+                    node.remove(aabb);
+                }
+            }
+        }
+
+        public boolean isLeaf () {
+            return children == null;
+        }
+
+        private void addGeometry (T geometry) {
+            if (geometries == null) {
+                geometries = new ObjectSet<>();
+            }
+            geometries.add(geometry);
+        }
+
+        public ObjectSet<T> query (BoundingBox aabb, ObjectSet<T> result) {
+            if (!isLeaf()) {
+                for (OctreeNode node : children) {
+                    node.query(aabb, result);
+                }
+            }
+            if (aabb.contains(bounds)) {
+                if (geometries != null) {
+                    result.addAll(geometries);
+                }
+            }
+            return result;
```

Comment:
else if aabb intersect with bounds, we need to add some geometries using "collider" to filter them

Reply:
I did it in a previous implementation. Ideally, collider should also be capable of filtering by frustum so they can have a similar behavior, what do you think?

Reply:
Added a method to the collider, makes sense to have it there.